Mathlib/LinearAlgebra/Matrix/SVD/Defs.lean

```diff
@@ -0,0 +1,479 @@
+  funext i j
+  cases' j with j j <;>
+  simp only [Equiv.refl_symm, Equiv.coe_refl, submatrix_apply, id_eq,
+    reindex_apply, of_apply, Sum.elim_inl, Sum.elim_inr]
+
+lemma reduced_spectral_theorem (A : Matrix (Fin M) (Fin N) 𝕂) :
+    Aᴴ  *  A = A.svdV₁  *  (A.svdμ.map (algebraMap ℝ 𝕂))  *  A.svdV₁ᴴ := by
+  let hAHA := isHermitian_transpose_mul_self A
+  -- "Ugly" (submatrix_mul_equiv) explicit rewrites: each one on its own line for
+  -- readability!!
+  rw [← submatrix_id_id (Aᴴ * A), IsHermitian.spectral_theorem' hAHA,
+    ← IsHermitian.conjTranspose_eigenvectorMatrix, Matrix.mul_assoc,
+    ← submatrix_mul_equiv
+      hAHA.eigenvectorMatrix (diagonal (IsROrC.ofReal ∘ hAHA.eigenvalues)  *
+      (hAHA.eigenvectorMatrixᴴ)) _ (eigenColumnEquiv A).symm _,
+    ← submatrix_mul_equiv (diagonal (IsROrC.ofReal ∘ hAHA.eigenvalues)) (hAHA.eigenvectorMatrixᴴ)
+    _ (eigenColumnEquiv A).symm _,
+    ← @IsROrC.algebraMap_eq_ofReal 𝕂]
+  simp_rw [Function.comp]
+  rw [← diagonal_map, submatrix_map, ← reindex_apply, ← Equiv.coe_refl, ← Equiv.refl_symm,
+    ← reindex_apply, ←conjTranspose_submatrix, ← reindex_apply, S_block, V_columns,
+    conjTranspose_fromColumns_eq_fromRows_conjTranspose, fromBlocks_map, fromBlocks_mul_fromRows,
+    fromColumns_mul_fromRows]
+  simp only [map_zero, Matrix.map_zero, Matrix.zero_mul, add_zero, Matrix.mul_zero]
+  rw [Matrix.mul_assoc]
+  apply map_zero
+
+lemma reduced_spectral_theorem' (A : Matrix (Fin M) (Fin N) 𝕂) :
+    A  *  Aᴴ = A.svdU₁'  *  (A.svdμ'.map (algebraMap ℝ 𝕂))  *  A.svdU₁'ᴴ := by
+  let hAAH := isHermitian_mul_conjTranspose_self A
+  rw [← submatrix_id_id (A * Aᴴ), IsHermitian.spectral_theorem' hAAH,
+    ← IsHermitian.conjTranspose_eigenvectorMatrix, Matrix.mul_assoc,
+    ← submatrix_mul_equiv hAAH.eigenvectorMatrix
+      (diagonal (IsROrC.ofReal ∘ hAAH.eigenvalues)  *  (hAAH.eigenvectorMatrixᴴ)) _
+        (eigenRowEquiv A).symm _,
+    ← submatrix_mul_equiv (diagonal (IsROrC.ofReal ∘ hAAH.eigenvalues))
+      (hAAH.eigenvectorMatrixᴴ) _ (eigenRowEquiv A).symm _,
+      ← @IsROrC.algebraMap_eq_ofReal 𝕂]
+  simp_rw [Function.comp]
+  rw [← diagonal_map, submatrix_map,
+    ← reindex_apply, ← Equiv.coe_refl, ← Equiv.refl_symm, ← reindex_apply,
+    ← conjTranspose_submatrix, ← reindex_apply, S'_block, U_columns',
+    conjTranspose_fromColumns_eq_fromRows_conjTranspose, fromBlocks_map,
+    fromBlocks_mul_fromRows, fromColumns_mul_fromRows]
+  simp only [map_zero, Matrix.map_zero, Matrix.zero_mul, add_zero, Matrix.mul_zero]
+  rw [Matrix.mul_assoc]
+  rw [map_zero]
+
+open scoped ComplexOrder
+lemma eig_vals_ne_zero_pos {m n: Type} [Fintype m] [Fintype n] [DecidableEq n]
+    (A : Matrix m n 𝕂) (z: {a // (isHermitian_transpose_mul_self A).eigenvalues a ≠ 0 }) :
+    0 < ((isHermitian_transpose_mul_self A).eigenvalues z) :=
+  lt_of_le_of_ne
+    (Matrix.PosSemidef.eigenvalues_nonneg (posSemidef_conjTranspose_mul_self _) _) -- 0 ≤ _
+    (z.prop.symm) -- 0 ≠ _
```

Comment:
It looks like you're using this result in two forms: 3 times showing the eigenvalue is not zero (but that's true by definition of `z`!) and 1 time showing that the eigenvalue is nonnegative. So I'd simply show that all eigenvalues of `A^H * A` are nonnegative, and rename to `eigenvalues_transpose_mul_self_nonneg`.

With those modifications, this belongs in an earlier file, probably soon after `Matrix.PosSemidef.eigenvalues_nonneg`.

Reply:
Than would be the direct combination of  [Matrix.PosSemidef.eigenvalues_nonneg](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/Matrix/PosDef.html#Matrix.PosSemidef.eigenvalues_nonneg) and [Matrix.posSemidef_conjTranspose_mul_self](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/Matrix/PosDef.html#Matrix.posSemidef_conjTranspose_mul_self).

Reply:
I mean it is exactly the following:

```lean
lemma eig_vals_ne_zero_pos' (A : Matrix (Fin M) (Fin N) 𝕂) : 
    ∀ i, 0 ≤ ((isHermitian_transpose_mul_self A).eigenvalues i) := 
  Matrix.PosSemidef.eigenvalues_nonneg (Matrix.posSemidef_conjTranspose_mul_self _)
```